### PR TITLE
Torchsparse

### PR DESCRIPTION
**Pull request by nicolas-chaulet**
Based on our conversation on Slack, here is a first proposal. What do you think?
Main downside is that the only way to set the backend is to use an environment variable. Any idea about how to make that a little easier? Like a set_backend method?

Also, we can grow the list of methods and modules, just wanted to start somewhere

### CODE REVIEW

**Review by humanpose1 — COMMENTED** [review]
Overall looks great ! 
The advantage of your implementation is that with few line of code, we can switch from Torchsparse to Minkowski.
But I'm not very fond of the `os.environ`. It would be better if we can add it in the yaml.
About the backend, I think it would be better if it is an option of each High level blocks (ResBlockDown and ResBlockUp).
It will allow to specify the backend easily in the yaml.

About the tests, I was thinking 
- basic tests to see that everythings works
- we can also add the code for benchmarking both backend 
- weight compatibility
To see if Minkowski weights and torchsparse weights are compatible, we can do something like
```python
class TestCase(unittest.TestCase):
    def test_weight_compatibility(self):
         model1 = Unet(backend="minkowski")
         model2 = Unet(backend="torchsparse")
         w1 = model1.state_dict()
         model2.load_state_dict_with_same_shape(w1)
         for k, p in w2.items():
             torch.testing.assert_allclose(w1[k], p)
         
```

This is my suggestion.
What do you think ?

File: torch_points3d/models/segmentation/sparseconv.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, option, model_type, dataset, modules):
+        # call the initialization method of UnetBasedModel
+        super().__init__(option, model_type, dataset, modules)
+        self.loss_names = ["loss_seg"]
+        self.head = nn.Sequential(nn.Linear(option.up_conv.up_conv_nn[-1][-1], dataset.num_classes))
+
+    def set_input(self, data, device):
+        self.raw_input = data
+        coords = torch.cat([data.coords.int(), data.batch.unsqueeze(-1).int()], -1)
+        self.input = SparseTensor(data.x, coords, device)
```
[the remark this answers is not in the document]
Comment (nicolas-chaulet) [reply torch_points3d/models/segmentation/sparseconv.py:26]: the `create_sparse_tensor` is actually handled in each implementation separately (see `SparseTensor` function) and handles each case separately

File: conf/models/segmentation/sparseconv.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+models:
+  ResUNet32:
+    class: sparseconv.Unet
```
[the remark this answers is not in the document]
Comment (nicolas-chaulet) [reply conf/models/segmentation/sparseconv.yaml:3]: Yes will add

**Comment by nicolas-chaulet** [comment]
@humanpose1 yes valid points. Regarding the backbend option, I think it should really be hidden so that only people who know what they are doing need to worry about that. It is still quite easy to switch backend, see tests as an example. I will clean that up later today and add a new API that supports both backends. Also, weights can be exchanged it seems :)

**Review by humanpose1 — COMMENTED** [review]
Great work !! It looks great ! When you finish I will adapt the code for registration and we'll be good I think ;).

**Comment by nicolas-chaulet** [comment]
@humanpose1 do you want to change your things in a next PR? Minkowski API is still up, I have just added a new one
